### ISSUE TITLE
Use a YAML file as the basis for the change log

### ISSUE THREAD

**Issue by taldcroft** [astropy/astropy#10334]
### Description

The repo at https://github.com/taldcroft/astropy-changelog shows a proof of concept for changing the basis of the astropy changelog from the `CHANGES.rst` file to a new `CHANGES.yml` file.  Please see the README there for details.  The driver for this and initial discussions were captured in an astropy-dev email thread in Oct 2016 (subject "Proposal for a new changelog system"). 

This was discussed in more detail at the 2019 coordination meeting in Socorro, with the agreement at the time that @bsipocz and @Cadair would develop a different system that uses information embedded in the PR description and/or GitHub metadata. While arguably a better solution, 6 months have passed with no progress, and in fact the details of the better solution were not captured in any written form. @bsipocz has stated in private conversation that she does not have time to spend on this given many other higher priorities and is OK with the YAML solution.

I propose moving forward on the YAML solution unless @Cadair can commit to developing the better solution in the next few months. All details of the YAML solution and how it gets implemented (both code and process) are certainly open for discussion here.

The problems in the current CHANGES.rst implementation are real and persist. I personally had to rebase at least 5 times due to a changelog conflict in the last few weeks (including once literally ten minutes after the last rebase), and now I see the 5-day old #10296 is already blocked from merge because of a changelog conflict. I also made a mistake in the placement of a change entry. 

### Additional context

https://github.com/taldcroft/astropy-changelog

**Comment by mhvk** [comment]
:+1: to solving this before the next release, and also :+1: for using the `yaml` file by default.

That said, I still have some preference to avoid a roll-our-own approach, and think it might be better to just copy what numpy does (though they limit it to "what's new" like items, relying on the bug label to collect those from PRs directly, but there is nothing intrinsic that prevents using it generally); see https://github.com/numpy/numpy/blob/master/doc/release/upcoming_changes/README.rst, item 5 in the [numpy developer instructions](https://numpy.org/doc/stable/dev/index.html) and their [pyproject.toml](https://github.com/numpy/numpy/blob/master/pyproject.toml) for setup.

That said, IIRC, your "homebrew" approach was not exactly super-complicated, i.e., I should not overly worry about maintainability.

**Comment by pllim** [comment]
Would you be willing to implement this at a small scale as a proof of concept? It is hard for me to see how this really works without a practical example. We need to support at least these scenarios:

1. New major release.
2. Backport to bug fix of latest major release.
3. Backport to bug fix of LTS release (that we support for 2 years).

**Comment by mhvk** [comment]
@pllim - was this directed at me? @taldcroft has a POC on his branch already.

**Comment by pllim** [comment]
I am not sure who it was directed to. 😅 

I am more interested to see how it really looks over here with some real entries we already have. 🤷

**Comment by taldcroft** [comment]
@pllim - the proof of concept (POC) https://github.com/taldcroft/astropy-changelog has the real existing astropy changelog represented as YAML. I'm not sure if that is what you are looking for.

**Comment by taldcroft** [comment]
Agreed on trying to avoid inventing something new.

About using towncrier ala numpy and sunpy, the biggest obstacle I see is the lack of subpackage. This is a key level of granularity that is really a requirement for astropy IMO. So what would it take to have `<PR>.<subpackage>.<entry_type>.rst`?  From what I can see in the rather terse towncrier documentation, this is not possible.

Smaller niggles:
- Putting thousands of temporary files into the repo history with the intent of deleting them. Maybe that's one for the philosophers.
- It is not common, but does happen that there are two or more PR's that are associated with a single change.  This is mostly very large features that have follow-on PR's, but towncrier cannot properly handle this (to my knowledge).
- I like the idea of having a structured YAML database of changes as a concept, though I have to admit I'm not immediately sure what I would do with this database.

**Comment by mhvk** [comment]
Ah, good point about the subpackages. I think that can be solved somewhat by entering fragments as something like `coordinates-12345.bugfix.rst` - this will order them correctly (but not give a link to github).

Anyway, since I don't think I'll be the one doing this, and do think your scheme is sensible, I'm quite happy to just go with that.

**Comment by pllim** [comment]
tl;dr -- Are we going to backport this new format to all the old releases or only for going forward?

**Comment by pllim** [comment]
As far as I can find on the POC repo, the workflow for maintainer is still undetermined, as well as the cadence of YAML to RST generation.

**Comment by taldcroft** [comment]
> As far as I can find on the POC repo, the workflow for maintainer is still undetermined, as well as the cadence of YAML to RST generation.

Indeed these are the key unresolved issues.

**Comment by taldcroft** [comment]
I thought some more and looked a little at towncrier.  I think adding support for subsections might not be too difficult (famous last words).  See : https://github.com/twisted/towncrier/issues/230.

**Comment by mhvk** [comment]
I also had a quick look and it indeed didn't seem too hard - even just having more flexibility in the format string would help. If the towncrier folks are interested, I'm happy to help.

**Comment by taldcroft** [comment]
@bsipocz - are you OK with a towncrier solution if that becomes possible?

**Comment by pllim** [comment]
Some follow-up questions:

* How long do we wait for replies from `towncrier` devs before giving up?
* Over at #10286 , there's a request to make API mentions live links. Is that possible with either Tom A's or towncrier solutions? We removed the links because they tend to break when API changes years down the road, but users find lack of linking hinders discoverability.

**Comment by taldcroft** [comment]
Good question about `towncrier`.  It does not give me a warm feeling that they are completely unresponsive to someone volunteering to contribute.

About live links, the YAML solution does give an opportunity to check links on-the-fly because the entire CHANGES.rst can be regenerated at any time from the source YAML.  With towncrier you lose the source information with each release and have only the rendered RST.

**Comment by embray** [comment]
I agree in principle with all of @taldcroft's points [here](https://github.com/astropy/astropy/issues/10334#issuecomment-626309741).  

However, perhaps you're already aware, CPython has been using towncrier successfully for [several years now](https://github.com/python/cpython/pull/552), so I think that if it can work for a project the size of CPython, it can work for Astropy.  
CPython also has a PR check that automatically checks if a changelog entry has been completed (there is an associated label maintainers can add for PRs that don't require a changelog entry).  This is all infrastructure we could probably borrow from CPython.

If there are features we need to add to towncrier and upstream is not responsive enough, there's no strong reason we can't use our own fork of it until/unless we have what we need.  Perhaps it's an annoyance to maintain a fork, but once it's working it's something that won't need to be updated often.

**Comment by saimn** [comment]
CPython is using blurb, which seems similar to towncrier but developed for their own needs: https://pypi.org/project/blurb/, https://devguide.python.org/committing/?highlight=blurb#what-s-new-and-news-entries
They also have a bot that can commit a news entry based on the PR (title, I guess): https://github.com/apps/blurb-it

One thing I like about towncrier, and I guess it would work the same with blurb, is that news fragments go the each release branch. So finally it would be possible to have news entries for bugfixes releases also listed in the feature releases. Currently if you look at e.g. 4.1 bug fixes, there is almost nothing, because most bugfix entries go only to 4.0.2. For users it is then easy to miss which bugs were fixed in a given feature release, but more annoying is when digging though the history to find in which feature release a bug was fixed. If you look at 3.1 or 3.2 bugfixes, the list is incomplete. How to find the other ones? Currently you have to find the closest in time 2.0.x release.

**Comment by taldcroft** [comment]
I'm not fighting that hard for the YAML solution any more, but I still don't see that the current towncrier can meet our need to have changes divided by sub-package. I essentially volunteered to add this functionality in https://github.com/twisted/towncrier/issues/230 and got zero response even after a 2nd ping. That sort of behavior in a project makes me quite hesitant to contribute. I suppose we could fork towncrier, but that seems to be missing the point even worse.

So we are left with:
- Status quo, which has worked for 10 years but has real deficiencies
- Use the YAML idea, which is a custom homebrew but has the advantage of being mostly implemented.
- Update towncrier to suit our needs and hope the change is accepted, or else fork if not.
- Use towncrier and stop dividing changes by sub-package (I'm strongly :-1: on that, however)
- Use towncrier and put in some other ad-hoc sub-package indicator that kinda works

**Comment by taldcroft** [comment]
@saimn - it's worth noting that the issue you mention about getting bug fixes listed in each of the relevant releases is handled with the YAML version.

In this concept the targeted versions for a particular PR are put into the YAML entry by the PR developr and/or sub-package maintainer prior to merge. This relieves the astropy release manager of the responsibility for knowing which release branches to put the PR into. To put another way, it makes the list of target branches for a PR be an explicitly documented list in a version-controlled file instead of an implicit list based on a single tag in the GitHub PR.

**Comment by saimn** [comment]
Thanks for the clarification @taldcroft, I meant to ask but forgot.
I agree with your status above, it would be great to make a decision. It's a pity that towncrier folks didn't answer, it seems there is no activity on the repo since May.
Still I think towncrier would be the best solution - if we find a solution for the sub-packages - as the YAML solution still has some of the drawbacks we face currently: 
- one huge file, which can be intimidating for new contributors, and subject to conflicts (there may be some git merge strategy that could help?). 
- the need to wait for a milestone to be set, and to update if the milestone is modified.
The workflow of towncrier/blurb makes things much easier both for the PR author and the maintainer / release manager.

**Comment by embray** [comment]
@saimn Thanks for the correction.  I forgot that it actually ended up being blurb.  I've interacted with it before via PRs and it's the same idea though.

**Comment by bsipocz** [comment]
> One thing I like about towncrier, and I guess it would work the same with blurb, is that news fragments go the each release branch. So finally it would be possible to have news entries for bugfixes releases also listed in the feature releases. Currently if you look at e.g. 4.1 bug fixes, there is almost nothing, because most bugfix entries go only to 4.0.2. For users it is then easy to miss which bugs were fixed in a given feature release, but more annoying is when digging though the history to find in which feature release a bug was fixed. If you look at 3.1 or 3.2 bugfixes, the list is incomplete. How to find the other ones? Currently you have to find the closest in time 2.0.x release.

I always saw this behaviour of not listing things twice, as a feature rather than a bug. Arguably releases happen in a serial time frame rather than parallel, e.g. 4.0.2 happens before 4.1. With this logic should all bugfixes from all the 4.0.x also be listed for 4.1?

Having the entries twice would certainly confuse me as I don't think any of the other projects where I regularly interact with the changelog work that way.

But, if this doubling behaviour is the wish of the majority, the switch with the current system is very easy. 

(although that case I would suggest a tiny bit different approach, namely that we just stop listing parallel version changelogs, changelogs stay on their bugfix branches. E.g. LTS releases would never appear in the latest branch (e.g. 2.0.1x would appear on a 3.2 as those entries was listed somewhere already in a 3.1, etc release).

**Comment by mhvk** [comment]
From a quick look, blurb does seem to have support for sections (like `Library` for CPython).

Getting rid of merge conflicts would still be fantastic, whatever way we choose!

**Comment by bsipocz** [comment]
> one huge file, which can be intimidating for new contributors

is there any data on this? 

also, is there any data on how helpful it is that one only needs to look into one file? (anecdotally, rather than data driven I certainly find it much much easier when I need to dig into a dev version of a package to look for a fix if they have everything already a single file in the final changelog's format, rather than they have gazillion of tiny files with no milestones attached).

**Comment by mhvk** [comment]
Agreed with @bsipocz that a big file, if anything, has advantages, in giving immediate examples of the kind of entries one would like to make.

I thought the main reason for changing things was the fact that currently we run into these annoying merge conflicts (and they are very annoying!).

**Comment by bsipocz** [comment]
merge conflicts could be resolved/minimized:
-  with a github action rebase bot, possibly, but none on the devops team has sufficient github access permissions for the repo to play with these options.
- having less last minute rush before a feature freeze that cause most of the conflicts. But as it turned out, this is a heretic aspiration. Given the periodic burst pattern served as a pride and bragging point for a few I have my doubts, but maybe now that some funding is involved things change.

**Comment by saimn** [comment]
> I always saw this behaviour of not listing things twice, as a feature rather than a bug. Arguably releases happen in a serial time frame rather than parallel, e.g. 4.0.2 happens before 4.1. With this logic should all bugfixes from all the 4.0.x also be listed for 4.1?

Yes. If you look at some release, e.g. 3.0/3.1/3.2, there are some bug fixes, but not all the bug fixes that were backported to a 2.0.x release. And releases are not ordered by date, so if you want to find all of them, first you have to know that some are missing, and then you must find the closest in time 2.0.x release. I don't see how this can be a feature.

**Comment by saimn** [comment]
So, I had a look at towncrier's code, and they actually a "section" feature that is exactly what we need, and what @taldcroft had requested in his ticket. But it's not documented ... https://github.com/twisted/towncrier/issues/61

With this feature : 
```
~/dev/astropy master !1 ?1                                                                    
❯ tree docs/changes
docs/changes
├── config
│   └── 145.bugfix
└── io.fits
    └── 123.feature

2 directories, 2 files
~/dev/astropy master !1 ?1                                                                    
❯ towncrier build --draft --version 5.0
Loading template...
Finding news fragments...
Rendering news fragments...
Draft only -- nothing has been written.
What is seen below is what would be written.

5.0 (2020-09-10)
================
astropy.config
--------------

Bug Fixes
^^^^^^^^^

- Add your info here (#145)


astropy.io.fits
---------------

New Features
^^^^^^^^^^^^

- Add your info here (#123)
```
The hierarchy of "sections" and "fragments" is inverted compared to what we have currently, but there is a rst template for this that can be customized.

**Comment by bsipocz** [comment]
> I don't see how this can be a feature.

I see it as consistency, that things are listed at only one place, and the date of that place what shows when it's released. But now that you highlighted it (as I recall this was never ever mentioned before as an issue, certainly not as a motivation to switch over to a totally different system).

I certainly would see the benefit and logic of having two separate changelogs, basically cutting off the LTS  and stable releases and duplicating bug entries between the two.

**Comment by taldcroft** [comment]
Some more points...

A big file is a good thing from the perspective of merge conflicts because if you insert a new entry based on any quasi-random process then merge conflicts will be quite rare. 

I really don't think that there are many people that manage to get through the long and rather intimidating documentation we have for [making a code contribution](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html) but are scared off by opening a big file.

AFAIK towncrier still doesn't handle the case of a news item that relates to two or more PR's. In particular very large PR's sometimes have follow-up PR's which have been split out in the interest of keeping things manageable. Or other reasons, but if you go through the historical change log this is not especially rare.

If towncrier is really abandoned then we should be especially wary of adopting it IMO.

**Comment by taldcroft** [comment]
> I certainly would see the benefit and logic of having two separate changelogs, basically cutting off the LTS and stable releases and duplicating bug entries between the two.

Without getting into the details, this points to the utility in separating out the fundamental input data (the change entries) from the rendered output (the change log that people see).

The YAML idea makes this a clean separation and we can decide at any time to change our mind about the rendered output. Using towncrier you lose the original input data with each release and the only thing left are the rendered outputs.

**Comment by embray** [comment]
What about a middle ground, where we can organize changelog entries in Tom's YAML format (which I do quite like) but update it to allow combining from multiple files.  There can be one big master CHANGES.yml that contains, e.g. all the release information.  But still allow merging in entries from a whole directory of smaller YAML files (a CHANGES.d if you will).  How exactly it would be broken into multiple files is up for discussion, but an advantage to that approach is it wouldn't really matter, and we could adjust with experience.

This would solve problems with merge conflicts for example by writing a single `CHANGES.d/<issue-number>.yml` file as part of a PR.

This is not the same as the towncrier approach: the files would never be deleted.  It's just breaking one long YAML file into several which can be split up in any manner that makes sense at the time.

**Comment by taldcroft** [comment]
@embray - that works for me! :smile:

**Comment by pllim** [comment]
If we can stop rendering the change log too in the process, it is gonna save some doc build time.

**Comment by mhvk** [comment]
I like that approach too - best of both worlds. And it leaves it even easier to eventually provide a skeleton entry automatically, maybe even based on the commit message

**Comment by taldcroft** [comment]
There is the slight problem of the thousands of tiny change files that accumulate (about 2400 as of last Oct 2019). Maybe we don't care?

**Comment by pllim** [comment]
> thousands of tiny change files

The way Git works, a lot of small files but only one of them is changing at a time is better than one giant file that changes all the time, right?

**Comment by taldcroft** [comment]
I am not worried about git, but rather command line `ls` or navigating to that directory in GitHub.

**Comment by bsipocz** [comment]
I can't help feeling that this is converging to the combination of the worst from both approaches. I fail to see having the many files in subdirs will be much easier to irregular contributors than the current changelog, while adding the YAML structure would also add complexity (a reminder that the yaml format caused issues e.g. at the scipy paper, most contributors had issues of submitting a properly formatted one, even though there was a template).

 And it certainly won't be convenient for looking for a summary of recent changes either, basically checking things on github is out of the question one can only do the command line.

**Comment by taldcroft** [comment]
What about going back to the single file model, with a utility script in the astropy git repo (but not in the package) that people can run. This would do a few things:
- Query interactively for the relevant inputs
- Do up-front validation, e.g. are the sub-packages allowed?
- Automatically get the next PR number if not provided
- Generate and put the YAML at a random place in the file, thus making merge conflicts exceedingly unlikely.

**Comment by saimn** [comment]
In other related news, it seems towncrier now have an active maintainer and will have a new release hopefully soon...

**Comment by embray** [comment]
I should clarify, I did not suggest breaking the existing changelog into a million tiny files; that would be pointless.  Just that it should be possible to concatenate multiple YAML files together automatically to build the full changelog database, which would enable keeping entries for new PRs in single files if desired.  If they ever build up to a large number an occasional maintenance task can be to cat them all into the "main" file and delete them, but it doesn't have to be done every time either.

I also agree that a utility script for listing / filtering recent changes would be extremely useful regardless.  It would be useful with or without allowing multiple files.

**Comment by embray** [comment]
In any case, just a few weeks back on the project I'm already tired of dealing with minor changelog formatting issues every time one of my PRs lags.  I'll work on integrating Tom's solution into a PR and hopefully we can leave the bikeshedding for when problems with it actually occur.

**Comment by bsipocz** [comment]
I yet to see how to avoid tiring the people with the proper yaml formatting, as well as update all the scripts and bots to deal with the new format, while maintaining the readability of browsing the changelog in master in a humanly readable format. The yaml approach, as Simon mentioned above, still has the issue of updating the "formatting issues" when there is a milestone update on an outdated PR, etc. (or I'm not sure what other changelog formatting issues you are referring to).

**Comment by taldcroft** [comment]
@bsipocz - about updating all the scripts and bots, do you have an initial list of impacts? I expect that the code in those scripts will become simpler / more sane since they just need to read a nicely structured (machine-readable) data file instead of scraping a human-readable document.

As I mentioned, there is not a requirement that contributors generate the YAML by hand if you allow for a helper script in the source repo that does the job. So unless you reject that idea outright I don't see a reason to keep going with the "tiring people" with YAML objection. As a reminder such a script can also solve the "thousands of files" problem and the merge conflicts problem by putting all the YAML entries into a single file and placing it randomly. Yes, it's a new process for contributors, but arguably easier if there is a simple script that just asks questions and makes sure the answers are sensible.

I agree that not having convenient access to the formatted version of the changelog in master is a real issue. Could we have a GitHub action that generates the latest CHANGES.rst and commits to master when a PR gets merged to master? Seems possible and maybe not that hard from what I know.  @pllim?

**Comment by pllim** [comment]
> Could we have a GitHub action that generates the latest CHANGES.rst and commits to master when a PR gets merged to master?

Theoretically, yes. Whether the default token access is enough or not, I am not sure. If someone opens a PR to implement this new workflow, I can try to build that on top of the PR and see.

**Comment by bsipocz** [comment]
Ok, I give up. I don't see how a custom helper script will help casual contributors who find editing a text file cumbersome. After all there is a script to figure out the number for the next pr, yet it's always being complained about as an issue.
I neither see how it will help with milestone changes that I presume caused the annoyance @embray referred to. 
And parsing it with a new action is yet as another tiny piece of infrastructure while otherwise the home grown infrastructure is already attacked all the time that it should be cut back more.
So yes, the random order in yaml might help with conflicts (of it's random, but as humans are involved I would not be surprised to see edits mostly to the beginning or end of it causing more conflicts than we currently have), it's the same or only workaround for the rest of the things the changlog currently does.

**Comment by saimn** [comment]
One thing that is not clear to me with the YAML solution is the workflow with respect to milestones and backports. Having to put the milestone in the YAML file is annoying because it means that you have to wait for a maintainer to set the milestone, which often changes following discussions in a PR. So people will need to edit the YAML. And if the PR is to be backported, should PR authors specify multiple milestones ?
And then when backporting PRs, it happens that a PR cannot be backported, or that another PR is backported though it was not planned initially, so again the milestone needs to be updated (on the release branch ? then the YAML file will need to be merged back to master). 
All of this is not a problem with towncrier since the changelog fragment does not need the milestone.

**Comment by taldcroft** [comment]
@saimn - All of your comments are fair enough, but I would say that the YAML solution represents an improvement over the current process, which is the goal. 

Right now the contributor is implicitly matching the milestone by virtue of placement of the change text in CHANGES.rst (i.e. which section it goes in). As a reviewer I find this to be a pain because in order to check the changelog entry I need to typically open the file and scroll around to see that it is in the correct section. In the YAML version it is entirely obvious. All of the problems you mention are also problems of the current process, and made worse. E.g. if the milestone changes, right now the PR author has to move the text. In the YAML version the PR author just edits `4.2.1` => `4.3.0`. So to my mind that represents a small *improvement*.

The idea is for the YAML to explicitly specify all versions to which the PR will be backported. This puts the burden on the maintainer (at the time of approving) to ensure this is correct.  In the circumstance that the maintainer made a mistake and the backport cannot be completed then the release manager would indeed make a commit to master to fix the target releases for the PR. This seems simpler to me than the current process where the release manager needs to actually move the changelog entry to a new section in CHANGES.rst (for a failed backport).

To be clear, if towncrier is going to be supported (and hopefully by more than just one person), and somebody can show that it can fully support generating the current astropy CHANGES.rst, and work through the process changes, then I'm OK with that.  

I'm just not happy with the status quo, "do nothing" solution.

Onel comment on YAML vs towncrier. There is a philosophy difference in the approach. In YAML, the `changes.yml` file stores the record of changes in a structured database, and the changelog files are derived "reports" of the changes. The way we generate reports can be easily updated as needed. With towncrier you only get the generated report in one format and the underlying data are effectively thrown away at the time of report generation.

**Comment by taldcroft** [comment]
On the flip side, I do agree that towncrier is conceptually simpler with fewer moving parts.

**Comment by saimn** [comment]
@taldcroft - I agree with you on many points and would love to see a solution to this changelog thing. It's just that if we are doing the change, which in any case has a cost in term of changing habits and procedures and possibly writing helper scripts, I would prefer if we find a solution to this milestone issue.

> To be clear, if towncrier is going to be supported (and hopefully by more than just one person), and somebody can show that it can fully support generating the current astropy CHANGES.rst, and work through the process changes, then I'm OK with that.

Well, there is some activity on their repo, with two persons trying to clean some mess with their latest release and planning to do a new one. On the long term it's always hard to say, but probably not worse that many projects we depend on (last release of configobj is from 2014, pyyaml also had many years without real maintenance, etc.).

About supporting what we need with towncrier, I have shown in https://github.com/astropy/astropy/issues/10334#issuecomment-690619033 that it already can do what we need, and I also wrote a template to get the same format as we have now for the rendered changelog, so I would be happy to open a PR with that for more discussion.

> I'm just not happy with the status quo, "do nothing" solution.

:+1: 

> Onel comment on YAML vs towncrier. There is a philosophy difference in the approach. In YAML, the changes.yml file stores the record of changes in a structured database, and the changelog files are derived "reports" of the changes. The way we generate reports can be easily updated as needed. With towncrier you only get the generated report in one format and the underlying data are effectively thrown away at the time of report generation.

Indeed, but you have shown with your script that it's easy enough to parse the changelog if we need to convert it to something else ;), and I just used that (thanks!) to do some tests with towncrier.
